src/library/scala/concurrent/impl/Promise.scala

```diff
@@ -7,400 +7,441 @@
-     *  be called by transformWith.
-     */
-    protected[concurrent] final def linkRootOf(target: DefaultPromise[T]): Unit = link(target.compressedRoot())
-
-    /** Link this promise to another promise so that both promises share the same
-     *  externally-visible state. Depending on the current state of this promise, this
-     *  may involve different things. For example, any onComplete listeners will need
-     *  to be transferred.
-     *
-     *  If this promise is already completed, then the same effect as linking -
-     *  sharing the same completed value - is achieved by simply sending this
-     *  promise's result to the target promise.
+    /** Link this promise to the root of another promise.
      */
-    @tailrec
-    private def link(target: DefaultPromise[T]): Unit = if (this ne target) {
-      get() match {
-        case r: Try[_] =>
-          if (!target.tryComplete(r.asInstanceOf[Try[T]]))
+    @tailrec protected[concurrent] final def linkRootOf(target: DefaultPromise[T], link: Link[T]): Unit =
+      if (this ne target) {
+        val state = get()
+        if (state.isInstanceOf[Try[T]]) {
+          if(!target.tryComplete(state.asInstanceOf[Try[T]]))
             throw new IllegalStateException("Cannot link completed promises together")
-        case dp: DefaultPromise[_] =>
-          compressedRoot(dp).link(target)
-        case listeners: List[_] if compareAndSet(listeners, target) =>
-          if (listeners.nonEmpty)
-            listeners.asInstanceOf[List[CallbackRunnable[T]]].foreach(target.dispatchOrAddCallback(_))
-        case _ =>
-          link(target)
+        } else if (state.isInstanceOf[Link[T]]) {
+          // TODO: does relinking ever make sense?
+          state.asInstanceOf[Link[T]].relink(link = state.asInstanceOf[Link[T]], target = target, owner = this)
+        } else /*if (state.isInstanceOf[Callbacks[T]]) */ {
+          val l = if (link ne null) link else new Link(target)
+          val p = l.promise(this)
+          if (p ne this) {
+            if (compareAndSet(state, l)) {
+              if (state ne Noop) p.dispatchOrAddCallbacks(p.get(), state.asInstanceOf[Callbacks[T]])
+            } else linkRootOf(p, l)
+          }
+        }
       }
-    }
   }
 
-  /** An already completed Future is given its result at creation.
-   *
-   *  Useful in Future-composition when a value to contribute is already available.
+  // Constant byte tags for unpacking transformation function inputs or outputs
+  final val Xform_noop          = 0
+  final val Xform_map           = 1
+  final val Xform_flatMap       = 2
+  final val Xform_transform     = 3
+  final val Xform_transformWith = 4
+  final val Xform_foreach       = 5
+  final val Xform_onComplete    = 6
+  final val Xform_recover       = 7
+  final val Xform_recoverWith   = 8
+  final val Xform_filter        = 9
+  final val Xform_collect       = 10
+  //final val Xform_failed        = 11
```

Comment:
I assume having these as `0.toByte` etc. doesn't compile them to constants as efficiently?

Reply:
@NthPortal Yeah, it doesn't compile them to constants. :(

Reply:
My understanding is that ascribing the type to the declaration would still produce a literal ie:

final val Xform_noop: Byte = 0

Would be a constant byte literal. I could be wrong about that.

Reply:
It becomes a term of type `Byte`, defined with the literal "0". The inliner only inlines constant vals of literal types, which `Byte` isn't.

The fix for this was discussed within SIP-23 (aka 42.type) but it was dropped in the final SIP: https://docs.scala-lang.org/sips/42.type.html#byte-and-short-literals.

Reply:
Bummer. Thanks for the helpful info.